test/test_irods_http_api.py

```diff
@@ -1825,6 +1825,332 @@ def test_non_parallel_writes_with_data_exceeding_internal_write_threshold(self):
+            })
+            self.logger.debug(r.content)
+            self.assertEqual(r.status_code, 200)
+            result = r.json()
+            self.assertEqual(result['irods_response']['status_code'], 0)
+            self.assertEqual(result['checksum'][5:], checksum)
+
+        finally:
+            # Remove the data object.
+            r = requests.post(self.url_endpoint, headers=headers, data={
+                'op': 'remove',
+                'lpath': data_object,
+                'catalog-only': 0,
+                'no-trash': 1
+            })
+            self.logger.debug(r.content)
+
+    def execute_write_using_header_based_form(self, **args):
+        conn = http.client.HTTPConnection(config.test_config['host'], config.test_config['port'])
+
+        headers = {
+            'Authorization': f"Bearer {args['bearer_token']}",
+            'Content-Type': 'application/x-www-form-urlencoded',
+            'irods-api-request-op': args['fields']['op'],
+            'irods-api-request-offset': args['fields']['offset'],
+            'irods-api-request-parallel-write-handle': args['fields']['parallel-write-handle'],
+            'irods-api-request-stream-index': args['fields']['stream-index']
+        }
+        conn.request('POST', config.test_config['url_base'] + '/data-objects', bytes(args['bytes'], 'utf-8'), headers)
+
+        response = conn.getresponse()
+        result = json.loads(response.read().decode('utf-8'))
+        self.logger.debug(result)
+        conn.close()
+
+        self.assertEqual(response.status, 200)
+
+        return result
+
+    def test_parallel_writes_using_header_based_form(self):
+        headers = {'Authorization': 'Bearer ' + self.rodsuser_bearer_token}
+
+        # Indicates whether the parallel_write_shutdown operation needs to be called.
+        # This is used for clean up when errors occur.
+        invoke_parallel_write_shutdown = False
+
+        try:
+            # Tell the server we're about to do a parallel write.
+            data_object = os.path.join('/', self.zone_name, 'home', self.rodsuser_username, 'parallel_write_header_based_0.txt')
+            r = requests.post(self.url_endpoint, headers=headers, data={
+                'op': 'parallel_write_init',
+                'lpath': data_object,
+                'stream-count': 3
+            })
+            self.logger.debug(r.content)
+            self.assertEqual(r.status_code, 200)
+            result = r.json()
+            self.assertEqual(result['irods_response']['status_code'], 0)
+            parallel_write_handle = result['parallel_write_handle']
+
+            # Make sure the parallel_write_shutdown operation is invoked now that we
+            # have a parallel-write handle.
+            invoke_parallel_write_shutdown = True
+
+            # Write to the data object using the parallel write handle.
+            futures = []
+            with concurrent.futures.ThreadPoolExecutor(max_workers=3) as executor:
```

Comment:
Should this match `stream-count`, or does it just happen to be the same value?

Reply:
Yes, it should match `stream-count`. Will update that.

Reply:
Done.